Add Granulator UI script

Tsarpf opened the pull request:
Apologies in advance as a certain language model did most of the coding. It's been more than 10 years since I last used Lua, and I was just trying to get back to music making ASAP 😄 So the code is not great. But I think it's already quite useful as it is.

I thought I'd open the PR sooner rather than later to open it up for discussion and suggestions, before I forget about the whole thing.

Tsarpf commented:
One thing clearly missing is adding the ”catchup” functionality for the pots as used in rest of disting menus. Now their values jump when switching between setting the mean/std.

One option would be to change the grain delay and length to be set with the encoders not pots.

expertsleepers commented on ui_scripts/Granulator.lua:
```diff
@@ -0,0 +1,227 @@
+  end,
+
+  encoder1Push = function()
+    local r = getParameter(granulator, p_record)
```
getParameter() returns an integer. Comparing the result against 0.5 seems very inefficient.

expertsleepers commented on ui_scripts/Granulator.lua:
```diff
@@ -0,0 +1,227 @@
+local p_reverse, p_lfo_shape, p_grain_shape
+local p_grain_limit
+
+local delayMode = 1   -- 1: mean, 2: spread
```
If you're not planning on adding more modes, these could be booleans.

Tsarpf replied:
changed

expertsleepers commented on ui_scripts/Granulator.lua:
```diff
@@ -0,0 +1,227 @@
+      local newVal = current + whichWay * step
+      if newVal > 100 then newVal = 100 end
+      if newVal < -100 then newVal = -100 end
+      setParameter(granulator, p_lfo_depth, newVal)
```
setParameter() clamps to the valid range internally.

expertsleepers commented on ui_scripts/Granulator.lua:
```diff
@@ -0,0 +1,227 @@
+  button1Push = function()
+    button1Held = true
+    if button2Held then
+      local drone = getParameter(granulator, p_drone1_enable)
```
drone is an integer.

expertsleepers commented on ui_scripts/Granulator.lua:
```diff
@@ -0,0 +1,227 @@
+    local modeStr = "Delay:" .. delayStr .. " | Size:" .. sizeStr .. " | Pitch:" .. pitchStr .. " | LFO:" .. lfoStr
+    drawTinyText(10, 56, modeStr)
+    
+    local dM = math.floor(getParameter(granulator, p_delay_mean))
```
getParameter() returns integers - no need for math.floor()

expertsleepers reviewed (changes requested):
If you could also add something to the readme that lists all the scripts, that would be great.

Tsarpf commented:
- I added one more functionality, changing the dry/granulator gain via holding button 1/2 and turning the left encoder. Also added that it remembers the gain setting when you click the respective button to toggle on/off

- I was not sure what you meant by "_add something to the readme that lists all the scripts_". I assume you meant the README in `ui_scripts` not the root RADME? So I added this one and its parameters.
  - Probably adding what the buttons do to the README like I now did is not what you meant but this is just an idea. I could also move it to the top of the script as a comment, for example?

Tsarpf commented:
Ok the rendered markdown of the controls is not very readable with all the bullet points, will need to figure out some other formatting at least.

Tsarpf commented:
I only now realized I was missing the rate parameter completely, worked quite nicely without with rate just set at 5ms so I didn't notice. Also I want to add ability to change the spawn mode, probably with yet another button combo. Oh well.

And furthermore now there is a bug with toggling the drone enable also turning gains on/off, fixing...

expertsleepers commented on ui_scripts/README.md:
```diff
@@ -14,3 +14,28 @@ Example script from the user manual; controls an Augustus Loop.
 
 A simple example script, corresponding to the [LFO.json](../presets/LFO.json) factory preset.
+
+## Granulator.lua
```
Hyperlink, like the others, please.

expertsleepers commented on ui_scripts/README.md:
```diff
@@ -14,3 +14,28 @@ Example script from the user manual; controls an Augustus Loop.
+
+## Granulator.lua
+
+Pot1: Grain delay (=position) mean/spread
```
Please add a brief introductory line indicating what is required from the preset i.e. that there's a Granulator algorithm to control.

expertsleepers commented on ui_scripts/README.md:
```diff
@@ -14,3 +14,28 @@ Example script from the user manual; controls an Augustus Loop.
+Hold Button2 + turn Pot2: Rate mean
+Hold Button2 + turn Pot3: Rate spread
+Hold Button3 + press Button4: cycle Grain Shape
+HOld Button4 + press Button3: cycle Spawn Mode
```
Typo

Tsarpf commented:
Alright finally had some time, made those changes.

Tsarpf commented:
Oof the readme is quite unreadable now that I see it in the Github UI 😢

expertsleepers commented:
Feel free to reformat it!
